### PR TITLE
CMake: fix linking against libsodium statically

### PR DESCRIPTION
**Pull request by thesamesam**
Our previous processing would throw away all but the last
entry in the found STATIC_LIBRARIES, which in newer versions
of libsodium, might be "pthread" -- hence not trying to link
against libsodium at all.

We now generate a temporary list, mangle it, then return that.

Bug: https://bugs.gentoo.org/791031
Fixes: https://github.com/diasurgical/devilutionX/issues/2615
Signed-off-by: Sam James <sam@gentoo.org>

### CODE REVIEW

**Comment by thesamesam** [comment]
re CI failure: I'm not sure if devliutionX wants to support use of system libsodium (or any libraries) on Windows, but that's not my call to make, so I'll have it fall back to the old CMake module when pkgconfig can't be found for now.

**Comment by AJenbo** [comment]
Thanks for looking in to this.

> re CI failure: I'm not sure if devilutionX wants to support use of system libsodium (or any libraries) on Windows, but that's not my call to make, so I'll have it fall back to the old CMake module when pkgconfig can't be found for now.

Sounds good, 2 of the 3 Windows builds are done via cross compiling where libsodium is simply extracted to a folder in the system path.

**Comment by AJenbo** [comment]
> It's not clear why the CMake module started to fail,
> but it appears that it has a few bugs in how it handles
> fallbacks, and it wasn't ever official upstream anyway

It actually was, but they dropped it rather then maintain it.
https://github.com/jedisct1/libsodium/issues/929

File: CMakeLists.txt [hunk 1]
```diff
@@ -212,8 +212,19 @@ endif()
...
 if(NOT NONET)
   if(DEVILUTIONX_SYSTEM_LIBSODIUM)
-    set(sodium_USE_STATIC_LIBS ${DEVILUTIONX_STATIC_LIBSODIUM})
-    find_package(sodium REQUIRED)
+    # We really need to use pkgconfig to find libsodium.
+    # They don't seem to support the CMake module method anymore and
+    # pkgconfig is simpler anyway.
+    # https://github.com/diasurgical/devilutionX/issues/2615
+    find_package(PkgConfig)
+    if(PkgConfig_FOUND)
```
Comment (glebm) [inline CMakeLists.txt:220]: This doesn't make sense in the main CMakeLists.txt. Findsodium.cmake should be adjusted instead.
As far as I can tell, it already tries pkgconfig first but perhaps it does so incorrectly.

Comment (glebm) [reply CMakeLists.txt:220]: Note that I mean our own Findsodium.cmake here: https://github.com/diasurgical/devilutionX/blob/master/CMake/Findsodium.cmake

We do not use the upstream (now deleted) Findsodium.cmake

Comment (thesamesam) [reply CMakeLists.txt:220]: Thank for the pointers! On it.

**Comment by thesamesam** [comment]
@glebm @AJenbo Thanks for your help! Thanks to the advice, I've changed tactics a bit here. I think I've now found the "real" issue - the mangling of the list of static libraries was broken.

How're we looking now?

**Review by glebm — APPROVED** [review]
Looks good, thank you!

**Comment by escalade** [comment]
Well, this seems to have broken linking against libsodium dynamically for me :)

```
CMake Error at /storage/RetroELEC/build.RetroELEC-skylake.x86_64-9.2-devel/toolchain/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:218 (message):
  Could NOT find sodium (missing: sodium_LIBRARY_RELEASE
  sodium_LIBRARY_DEBUG) (found version "1.0.18")
Call Stack (most recent call first):
  /storage/toolchain/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
  CMake/Findsodium.cmake:236 (find_package_handle_standard_args)
  CMakeLists.txt:216 (find_package)
```

From the error log:
```
[2/2] Linking C executable cmTC_07440
FAILED: cmTC_07440 
: && /storage/toolchain/bin/x86_64-libreelec-linux-gnu-gcc -march=skylake -m64 -mmmx -msse -msse2 -mfpmath=sse -fomit-frame-pointer -Wall -pipe  -O2 -Wl,--sort-common -Wl,--as-needed -fuse-ld=gold -Wl,-O2 CMakeFiles/cmTC_07440.dir/src.c.o -o cmTC_07440   && :
CMakeFiles/cmTC_07440.dir/src.c.o:src.c:function main: error: undefined reference to 'pthread_create'
CMakeFiles/cmTC_07440.dir/src.c.o:src.c:function main: error: undefined reference to 'pthread_detach'
CMakeFiles/cmTC_07440.dir/src.c.o:src.c:function main: error: undefined reference to 'pthread_cancel'
CMakeFiles/cmTC_07440.dir/src.c.o:src.c:function main: error: undefined reference to 'pthread_join'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by AJenbo** [comment]
@escalade are you also using gentoo?

**Comment by thesamesam** [comment]
@escalade would you mind filing a new bug please and we’ll dig into it there? It’s be useful to have all the logs.

I can try setup which distro you’re using if needed.

**Comment by escalade** [comment]
Compiling with -DCMAKE_BUILD_TYPE=MinSizeRel rather than "Release" solved my issue. My toolchain is a heavily customized LibreELEC toolchain, which is similar to buildroot.

**Comment by thesamesam** [comment]
OK. I'm not sure why this would make a difference but if you hit it again, please do file a bug with all the details and I will happily look into it, of course! Glad it is OK now.

**Pull request merged**